### PR TITLE
rust: update toolchain to fix missing rls

### LINKED ISSUE #11953: Rls is missing
### What version of TiKV are you using?

master

### What operating system and CPU are you using?

No related

### Steps to reproduce

```
$ rustup component add rls                                                                
error: component 'rls' for target 'x86_64-unknown-linux-gnu' is unavailable for download for channel 'nightly-2022-01-07'
Sometimes not all components are available in any given nightly. If you don't need the component, you can remove it with:
```

### What did you expect?
Have a rls component

### PR DESCRIPTION
**Pull request by Connor1996**
### What is changed and how it works?

Issue Number: Close #11953 

What's Changed:

Update to rust nightly 2022-02-14 which is presented with rls, see https://rust-lang.github.io/rustup-components-history/

### Check List 

Tests 

- No code

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by tabokie — APPROVED** [review]

File: Cargo.lock [hunk 1]
```diff
@@ -6330,9 +6320,9 @@ dependencies = [
 
 [[package]]
 name = "x86"
-version = "0.37.0"
+version = "0.45.0"
```
Comment (BusyJay) [inline Cargo.lock:6323]: Why change this?

Comment (Connor1996) [reply Cargo.lock:6323]: `llvm_asm!` is replaced with `asm!` in recent nightly, so update the dependency to fix compile issue.

**Comment by 5kbpers** [comment]
Please fix the lint errors.

File: cmd/tikv-ctl/src/util.rs [hunk 2]
```diff
@@ -10,6 +10,7 @@ use tikv::config::TiKvConfig;
 
 const LOG_DIR: &str = "./ctl-engine-info-log";
 
+#[allow(clippy::field_reassign_with_default)]
```
Comment (tabokie) [inline cmd/tikv-ctl/src/util.rs:13]: This should be fixed.

Comment (Connor1996) [reply cmd/tikv-ctl/src/util.rs:13]: the original way is cleaner, and we also need to change some subfield of the field, so it needs be written in reassign anyway

**Comment by Connor1996** [comment]
PTAL

File: components/txn_types/src/timestamp.rs [hunk 3]
```diff
@@ -144,12 +145,14 @@ impl TsSet {
 
     pub fn from_u64s(ts: Vec<u64>) -> Self {
         // This conversion is safe because TimeStamp is a transparent wrapper over u64.
+        #[allow(clippy::unsound_collection_transmute)]
```
Comment (5kbpers) [inline components/txn_types/src/timestamp.rs:148]: This is undefined behavior. Better to fix this through iteration like `ts.into_iter().map(TimeStamp::from).collect()`

File: components/txn_types/src/timestamp.rs [hunk 4]
```diff
@@ -144,12 +145,14 @@ impl TsSet {
...
         let ts = unsafe { ::std::mem::transmute::<Vec<u64>, Vec<TimeStamp>>(ts) };
         Self::new(ts)
     }
 
     pub fn vec_from_u64s(ts: Vec<u64>) -> Self {
         // This conversion is safe because TimeStamp is a transparent wrapper over u64.
+        #[allow(clippy::unsound_collection_transmute)]
```
Comment (5kbpers) [inline components/txn_types/src/timestamp.rs:155]: (ditto)

File: components/concurrency_manager/src/lib.rs [hunk 5]
```diff
@@ -79,6 +79,7 @@ impl ConcurrencyManager {
             result[index] = MaybeUninit::new(self.lock_table.lock_key(key).await);
         }
         #[allow(clippy::unsound_collection_transmute)]
+        #[allow(clippy::transmute_undefined_repr)]
```
Comment (5kbpers) [inline components/concurrency_manager/src/lib.rs:82]: It seems also to be a UB

Comment (sticnarf) [reply components/concurrency_manager/src/lib.rs:82]: I don't understand. Why is it a UB to transmute from `Vec<MaybeUninit<T>>` to `Vec<T>`?

File: components/txn_types/src/timestamp.rs (hunk 3 above)
Comment (sticnarf) [reply components/txn_types/src/timestamp.rs:148]: I think it will be 100% safe if we add `#[repr(transparent)]` to `TimeStamp`.

Comment (Connor1996) [reply components/txn_types/src/timestamp.rs:148]: But `TimeStamp` is just a wrapper of u64. Seems these warnings are only due to clippy's check is more strict

File: components/concurrency_manager/src/lib.rs (hunk 5 above)
Comment (Connor1996) [reply components/concurrency_manager/src/lib.rs:82]: don't understand. Not sure if it's due to the clippy's check is more strict or the compile really changes the memory layout.

Comment (5kbpers) [reply components/concurrency_manager/src/lib.rs:82]: Ok I read the type of `result` wrong

File: components/txn_types/src/timestamp.rs (hunk 3 above)
Comment (Connor1996) [reply components/txn_types/src/timestamp.rs:148]: @sticnarf Tried it before, `repr(C)` eliminates `transmute_undefined_repr`, whereas `repr(transparent)` eliminates `unsound_collection_transmute`. There is always a clippy error anyway. 🤣

Comment (Connor1996) [reply components/txn_types/src/timestamp.rs:148]: Though `repr(transparent)` is more appropriate, will change it from `repr(C)`

File: components/concurrency_manager/src/lib.rs (hunk 5 above)
Comment (sticnarf) [reply components/concurrency_manager/src/lib.rs:82]: Ah, I know why it is UB. It is indeed possible that `Wrapper<MaybeUninit<T>>` may be different from `Wrapper<T>`. https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#layout has an example of `Option`.

Actually this function `lock_keys` is only used in tests and I don't believe `Vec` can have similar behavior as `Option` XD

Comment (Connor1996) [reply components/concurrency_manager/src/lib.rs:82]: got it

**Review by sticnarf — APPROVED** [review]
LGTM

**Comment by Connor1996** [comment]
PTAL @5kbpers

**Comment by sticnarf** [comment]
I see `test_read_index_retry_lock_checking` fails because of failing to transfer leader. But I can always pass the test at my dev env...

Looks strange...

**Pull request merged**